### PR TITLE
tests/gnrc_dhcpv6_client: Fix for newer Kea versions and remove sudo requirement

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
The fixes here are two-fold:

1. It fixes a regression, I believe was introduced in https://github.com/RIOT-OS/RIOT/pull/15771, where the DHCPv6 server could not be started, as it does not have root permissions when run with `make test-as-root`
2. It dynamically adapts the config JSON in that test using a small python script, to be able to run the DHCPv6 with Kea versions >= 1.7.10 (see https://kea.readthedocs.io/en/kea-1.7.10/arm/config.html#json-syntax for more details).

### Testing procedure
The tests should run without any adoptations on more up-to-date Linux distrubutions such as Arch Linux.

### Issues/PRs references
None

### CODE REVIEW

**Comment by miri64** [comment]
Might also be of interest for @JKRhb :-).

File: tests/gnrc_dhcpv6_client/Makefile [hunk 1]
```diff
@@ -39,10 +39,13 @@ TEST_ON_CI_BLACKLIST += all
 
 include $(RIOTBASE)/Makefile.include
 
-.PHONY: dhcpv6_server
+.PHONY: dhcpv6_server sudo
 
-dhcpv6_server:
-	$(CURDIR)/dhcpv6_server.sh $(DHCPV6_SERVER_PORT) $(CURDIR)/kea-dhcp6.conf
+dhcpv6_server: sudo
```
Comment (miri64) [inline tests/gnrc_dhcpv6_client/Makefile:44]: The `sudo` dependency is needed as the script goes directly to background, obscuring the password prompt. I am aware that this won't work on systems with sudo rules, where any sudo call requires a password input, but I think it is better than just letting the script in background fail due to missing sudo permissions.

**Comment by kfessel** [comment]
KEA_LOCKFILE_DIR=. KEA_PIDFILE_DIR=.

**Comment by kfessel** [comment]
KEA_LOCKFILE_DIR=. KEA_PIDFILE_DIR=. 

and a config that uses an unprevileged port  as well as keeping the database local  removes the need for kea beeing sudoed

    "lease-database": {
      "type": "memfile",
      "name": "kea-dhcp6.leases.csv"
    },

**Comment by miri64** [comment]
@kfessel does this also work for older versions of Kea?

**Comment by miri64** [comment]
Seems like it. The doc of Kea 1.6 at least also mentions these.

**Comment by miri64** [comment]
Mhhh but Debian and Ubuntu 18.04 use kea 1.1... Let me check with an older machine first

**Comment by miri64** [comment]
Mh did not find an older system I have quick access too, but both [KEA_LOCKFILE_DIR](https://gitlab.isc.org/isc-projects/kea/-/blob/de41bf1dd503a644496ff208b7a0d5503e12d235/src/lib/log/logger_impl.cc#L51) and [KEA_PIDFILE_DIR](https://gitlab.isc.org/isc-projects/kea/-/blob/de41bf1dd503a644496ff208b7a0d5503e12d235/src/lib/dhcpsrv/daemon.cc#L35) appear in the code base of [Kea v1.1.0](https://gitlab.isc.org/isc-projects/kea/-/tags/Kea-1.1.0). So I think it's safe to assume, they work.

**Comment by miri64** [comment]
Done as proposed in latest commit.

**Comment by miri64** [comment]
Don't know what happened, but when trying to rerun, there now is a different problem...:

```
2021-09-01 13:28:27.520 ERROR [kea-dhcp6.dhcp6/200173.140083192770240] DHCP6_INIT_FAIL failed to initialize Kea server: configuration error using file '/tmp/tmp.jLG3JlZCqY': unable to configure server identifier: unable to open DUID file /var/lib/kea/kea-dhcp6-serverid for writing
```

**Comment by kfessel** [comment]
"data-directory": ".", 

needs to be part of config

maybe mktemp -d 
cd there  and just have a config with these "."

**Comment by miri64** [comment]
An alternative could be to make the server-id not persistent (see https://kea.readthedocs.io/en/latest/arm/dhcp6-srv.html?highlight=duid%20storage#server-identifier-in-dhcpv6), but I think I like your solution better. It is definitely more future-proof :-).

**Comment by JKRhb** [comment]
> Might also be of interest for @JKRhb :-).

Very cool :) Should #16728 be rebased after this PR to include the tests in the CI?

**Comment by kfessel** [comment]
tested for the kea version less than  1.7.10 case

**Comment by miri64** [comment]
> Should #16728 be rebased after this PR has been merged to include the tests in the CI?

Your call if you want to rebase. Since this still is a `test-with-config` (tap interfaces are required) the CI won't run it anyways.

**Comment by miri64** [comment]
Fixed some shellcheck errors.

File: tests/gnrc_dhcpv6_client/dhcpv6_server.sh [hunk 2]
```diff
@@ -6,20 +6,55 @@
...
+    json.dump(config, c)
+EOF
+)"
+        python3 -c "$convert_json"
+    fi
     sleep 1 # sleep to let TAP become active
-    sudo kea-dhcp6 -p $1 -c $2
+    cd "$TMPDIR" || exit 1
+    KEA_PIDFILE_DIR=. KEA_LOCKFILE_DIR=. \
+        kea-dhcp6 -p "$1" -c "$config"
```
Comment (kfessel) [inline tests/gnrc_dhcpv6_client/dhcpv6_server.sh:56]:
```suggestion
    cd "$TMPDIR" || exit 1
    KEA_PIDFILE_DIR=. KEA_LOCKFILE_DIR=. \
        kea-dhcp6 -p "$1" -c "$config" &
```

File: tests/gnrc_dhcpv6_client/dhcpv6_server.sh [hunk 3]
```diff
@@ -6,20 +6,55 @@
...
 }
 
 # no need to kill from external, kea handles double instances gracefully
-_dhcpv6_server $1 $2 &
+_dhcpv6_server "$1" "$2" &
```
Comment (kfessel) [inline tests/gnrc_dhcpv6_client/dhcpv6_server.sh:60]:
```suggestion
_dhcpv6_server "$1" "$2"
```

**Review by kfessel — COMMENTED** [review]
we may as well leave the shell if its no longer used

**Comment by kfessel** [comment]
i got me a kea 1.8.0 -> tested

you may squash

**Comment by miri64** [comment]
> we may as well leave the shell if its no longer used

That does not work. The `sleep 1` needs to be called while the test is already running. Otherwise, on systems such as Arch, the TAP interface will not be connected yet, letting the test fail:

```
2021-09-01 15:56:26.164 INFO  [kea-dhcp6.dhcp6/229179.140037386780352] DHCP6_STARTING Kea DHCPv6 server version 1.9.10 (development) starting
2021-09-01 15:56:26.164 WARN  [kea-dhcp6.dhcp6/229179.140037386780352] DHCP6_DEVELOPMENT_VERSION This software is a development branch of Kea. It is not recommended for production use.
2021-09-01 15:56:26.165 INFO  [kea-dhcp6.hosts/229179.140037386780352] HOSTS_BACKENDS_REGISTERED the following host backend types are available: mysql postgresql 
2021-09-01 15:56:26.165 INFO  [kea-dhcp6.dhcpsrv/229179.140037386780352] DHCPSRV_CFGMGR_ADD_IFACE listening on interface tapbr0
2021-09-01 15:56:26.166 INFO  [kea-dhcp6.dhcpsrv/229179.140037386780352] DHCPSRV_CFGMGR_NEW_SUBNET6 a new subnet has been added to configuration: 2001:db8::/32 with params: t1=1000, t2=2000, preferred-lifetime=3000, valid-lifetime=4000, rapid-commit is false
2021-09-01 15:56:26.166 INFO  [kea-dhcp6.dhcp6/229179.140037386780352] DHCP6_CONFIG_COMPLETE DHCPv6 server has completed configuration: added IPv6 subnets: 1; DDNS: disabled
2021-09-01 15:56:26.166 INFO  [kea-dhcp6.dhcpsrv/229179.140037386780352] DHCPSRV_MEMFILE_DB opening memory file lease database: name=kea-dhcp6.leases.csv type=memfile universe=6
2021-09-01 15:56:26.166 INFO  [kea-dhcp6.dhcpsrv/229179.140037386780352] DHCPSRV_MEMFILE_LEASE_FILE_LOAD loading leases from file kea-dhcp6.leases.csv
2021-09-01 15:56:26.166 INFO  [kea-dhcp6.dhcpsrv/229179.140037386780352] DHCPSRV_MEMFILE_LFC_SETUP setting up the Lease File Cleanup interval to 3600 sec
2021-09-01 15:56:26.167 INFO  [kea-dhcp6.dhcp6/229179.140037386780352] DHCP6_USING_SERVERID server is using server-id 00:01:00:01:28:c2:34:1b:d4:6d:6d:fd:93:9b and stores in the file ./kea-dhcp6-serverid
2021-09-01 15:56:26.167 WARN  [kea-dhcp6.dhcpsrv/229179.140037386780352] DHCPSRV_OPEN_SOCKET_FAIL failed to open socket: the interface tapbr0 is not running
2021-09-01 15:56:26.167 WARN  [kea-dhcp6.dhcpsrv/229179.140037386780352] DHCPSRV_NO_SOCKETS_OPEN no interface configured to listen to DHCP traffic
2021-09-01 15:56:26.167 WARN  [kea-dhcp6.dhcp6/229179.140037386780352] DHCP6_MULTI_THREADING_INFO enabled: no, number of threads: 0, queue size: 0
2021-09-01 15:56:26.167 INFO  [kea-dhcp6.dhcp6/229179.140037386780352] DHCP6_STARTED Kea DHCPv6 server version 1.9.10 started
r
/home/mlenders/Repositories/RIOT-OS/RIOT/tests/gnrc_dhcpv6_client/bin/native/tests_gnrc_dhcpv6_client.elf tap0 
RIOT native interrupts/signals initialized.
LED_RED_OFF
LED_GREEN_ON
RIOT native board initialized.
RIOT native hardware initialization complete.

NETOPT_RX_END_IRQ not implemented by driver
Help: Press s to start test, r to print it is ready
READY
s
START
main(): This is RIOT! (Version: 2021.10-devel-364-g507d3-HEAD)
Iface  5  HWaddr: E2:BC:7D:CB:F5:50 
          L2-PDU:1500  MTU:1500  HL:64  Source address length: 6
          Link type: wired
          inet6 addr: fe80::e0bc:7dff:fecb:f550  scope: link  VAL
          inet6 group: ff02::1
          inet6 group: ff02::1:ffcb:f550
          
Iface  5  HWaddr: E2:BC:7D:CB:F5:50 
          L2-PDU:1500  MTU:1500  HL:64  Source address length: 6
          Link type: wired
          inet6 addr: fe80::e0bc:7dff:fecb:f550  scope: link  VAL
          inet6 group: ff02::1
          inet6 group: ff02::1:ffcb:f550
          
Timeout in expect script at "child.expect(r"inet6 addr:\s+(?P<global_addr>[0-9a-f:]+)\s+scope: global")" (tests/gnrc_dhcpv6_client/tests-with-config/01-run.py:59)
```

Note the lines

```
2021-09-01 15:56:26.167 WARN  [kea-dhcp6.dhcpsrv/229179.140037386780352] DHCPSRV_OPEN_SOCKET_FAIL failed to open socket: the interface tapbr0 is not running
2021-09-01 15:56:26.167 WARN  [kea-dhcp6.dhcpsrv/229179.140037386780352] DHCPSRV_NO_SOCKETS_OPEN no interface configured to listen to DHCP traffic
```

Without you suggestions it works fine.

File: tests/gnrc_dhcpv6_client/dhcpv6_server.sh (hunk 3 above)
Comment (kfessel) [reply tests/gnrc_dhcpv6_client/dhcpv6_server.sh:60]: so without this

File: tests/gnrc_dhcpv6_client/dhcpv6_server.sh (hunk 2 above)
Comment (kfessel) [reply tests/gnrc_dhcpv6_client/dhcpv6_server.sh:56]: but with this - which will leave the function after the server is kicked off

**Review by kfessel — APPROVED** [review]
:tada:

**Pull request merged**